src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+    pub fn with_complete_callback(&mut self, complete_cb: Job) -> Batch {
+        self.complete(complete_cb);
+        self.success(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with both `success` and `complete` callbacks specified.
+    pub fn with_callbacks(&mut self, success_cb: Job, complete_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(complete_cb);
+        self.build()
+    }
+}
+
+pub struct BatchHandle<'a, S: Read + Write> {
+    bid: String,
+    prod: &'a mut Producer<S>,
+}
+
+impl<'a, S: Read + Write> BatchHandle<'a, S> {
+    /// ID issued by the Faktory server to this batch.
+    pub fn id(&self) -> &str {
+        self.bid.as_ref()
+    }
+
+    pub(crate) fn new(bid: String, prod: &mut Producer<S>) -> BatchHandle<'_, S> {
+        BatchHandle { bid, prod }
+    }
+
+    /// Add the given job to the batch.
+    pub fn add(&mut self, mut job: Job) -> Result<(), Error> {
+        job.custom.insert("bid".into(), self.bid.clone().into());
+        self.prod.enqueue(job)
+    }
+
+    /// Initiate a child batch of jobs.
+    pub fn start_batch(&mut self, mut batch: Batch) -> Result<BatchHandle<'_, S>, Error> {
+        batch.parent_bid = Some(self.bid.clone());
+        self.prod.start_batch(batch)
+    }
+
+    /// Commit this batch.
+    ///
+    /// The Faktory server will not queue any callbacks, unless the batch is committed.
+    /// Committing an empty batch will make the server queue the callback(s) right away.
+    /// Once committed, the batch can still be re-opened with [open_batch](struct.Producer.html#method.open_batch),
+    /// and extra jobs can be added to it.
+    pub fn commit(self) -> Result<(), Error> {
+        self.prod.commit_batch(self.bid)
+    }
+}
+
+/// Batch status retrieved from Faktory server.
+#[derive(Deserialize, Debug)]
+pub struct BatchStatus {
+    // Fields "bid", "created_at", "description", "total", "pending", and "failed"
+    // are described in the docs: https://github.com/contribsys/faktory/wiki/Ent-Batches#status
+    /// Id of this batch.
+    pub bid: String,
+
+    /// Batch creation date and time.
+    pub created_at: DateTime<Utc>,
+
+    /// Batch description, if any.
+    pub description: Option<String>,
+
+    /// Number of jobs in this batch.
+    pub total: usize,
+
+    /// Number of pending jobs.
+    pub pending: usize,
+
+    /// Number of failed jobs.
+    pub failed: usize,
+
+    // The official golang client also mentions "parent_bid', "complete_st", and "success_st":
+    // https://github.com/contribsys/faktory/blob/main/client/batch.go#L8-L22
+    /// Id of the parent batch, provided this batch is a child ("nested") batch.
+    pub parent_bid: Option<String>,
+
+    /// State of the `complete` callback.
+    ///
+    /// See [complete](struct.Batch.html#structfield.complete).
+    #[serde(rename = "complete_st")]
+    pub complete_callback_state: String,
```

Comment:
Can we provide a more strongly typed value here than `String`?

Reply:
I've been thinking it over, since it's so tempting to enumerate those states, but I never found any promises on this part. Will check once again

Reply:
Confirmation from the Faktory's maintainer will be needed here. Emailed to Mike Perham and CC'd you.

Reply:
After receiving confirmation from the Faktory maintainers that they are not planning to change the API, enumerated those states via `CallbackState`, so we've got:
```rust
let status = tracker.get_batch_status(bid)?.unwrap();
match status.complete_callback_state {
    CallbackState::Pending => {},
    CallbackState::Enqueued => {},
    CallbackState::FinishedOk => {},
}
```
One can wonder, why there is no `FinshedNotOk` variant. I've updated one of the tests (`test_callback_will_be_queued_upon_commit_even_if_batch_is_empty`) to demonstrate that if the callback has been executed with a failure, it's state just stays `CallbackState::Enqueued`.